doxygenDocs/man/man3/CoolBoard.3
.TH "CoolBoard" 3 "Tue Aug 8 2017" "CoolAPI" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CoolBoard \- This class manages the \fBCoolBoard\fP and all of Its functions\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <CoolBoard\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBCoolBoard\fP ()"
.br
.ti -1c
.RI "void \fBbegin\fP ()"
.br
.ti -1c
.RI "bool \fBconfig\fP ()"
.br
.ti -1c
.RI "void \fBupdate\fP (const char *\fBanswer\fP)"
.br
.ti -1c
.RI "void \fBoffLineMode\fP ()"
.br
.ti -1c
.RI "void \fBonLineMode\fP ()"
.br
.ti -1c
.RI "int \fBconnect\fP ()"
.br
.ti -1c
.RI "int \fBisConnected\fP ()"
.br
.ti -1c
.RI "unsigned long \fBgetLogInterval\fP ()"
.br
.ti -1c
.RI "void \fBprintConf\fP ()"
.br
.ti -1c
.RI "void \fBsleep\fP (unsigned long interval)"
.br
.ti -1c
.RI "String \fBreadSensors\fP ()"
.br
.ti -1c
.RI "void \fBinitReadI2C\fP ()"
.br
.ti -1c
.RI "void \fBstopReadI2C\fP ()"
.br
.ti -1c
.RI "String \fBuserData\fP ()"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBCoolFileSystem\fP \fBfileSystem\fP"
.br
.ti -1c
.RI "\fBCoolBoardSensors\fP \fBcoolBoardSensors\fP"
.br
.ti -1c
.RI "\fBCoolBoardLed\fP \fBcoolBoardLed\fP"
.br
.ti -1c
.RI "\fBCoolTime\fP \fBrtc\fP"
.br
.ti -1c
.RI "\fBCoolWifi\fP \fBwifiManager\fP"
.br
.ti -1c
.RI "\fBCoolMQTT\fP \fBmqtt\fP"
.br
.ti -1c
.RI "\fBJetpack\fP \fBjetPack\fP"
.br
.ti -1c
.RI "\fBIrene3000\fP \fBirene3000\fP"
.br
.ti -1c
.RI "\fBExternalSensors\fP \fBexternalSensors\fP"
.br
.ti -1c
.RI "bool \fBuserActive\fP =0"
.br
.ti -1c
.RI "bool \fBireneActive\fP =0"
.br
.ti -1c
.RI "bool \fBjetpackActive\fP =0"
.br
.ti -1c
.RI "bool \fBexternalSensorsActive\fP =0"
.br
.ti -1c
.RI "bool \fBsleepActive\fP =0"
.br
.ti -1c
.RI "unsigned long \fBlogInterval\fP =1"
.br
.ti -1c
.RI "int \fBserverTimeOut\fP =180"
.br
.ti -1c
.RI "String \fBdata\fP =''"
.br
.ti -1c
.RI "String \fBanswer\fP =''"
.br
.ti -1c
.RI "const int \fBEnI2C\fP = 5"
.br
.in -1c
.SH "Detailed Description"
.PP 
This class manages the \fBCoolBoard\fP and all of Its functions\&. 
.PP
Definition at line 33 of file CoolBoard\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "CoolBoard::CoolBoard ()"
\fBCoolBoard::CoolBoard()\fP: This Constructor is provided to start the I2C interface and Init the different used pins 
.PP
Definition at line 34 of file CoolBoard\&.cpp\&.
.PP
References EnI2C\&.
.PP
.nf
35 {
36 
37 #if DEBUG == 1
38 
39     Serial\&.println( F("Entering CoolBoard Constructor") );
40     Serial\&.println();
41 
42 #endif
43     
44     Wire\&.begin(2, 14);                       //I2C init 
45 
46     pinMode(EnI2C, OUTPUT);        //Declare I2C Enable pin 
47 
48 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void CoolBoard::begin ()"
\fBCoolBoard::begin()\fP: This method is provided to configure and start the used CoolKit Parts\&. It also starts the first connection try If Serial is enabled,it prints the configuration of the used parts\&. 
.PP
Definition at line 59 of file CoolBoard\&.cpp\&.
.PP
References Jetpack::begin(), CoolMQTT::begin(), ExternalSensors::begin(), CoolBoardSensors::begin(), CoolWifi::begin(), CoolTime::begin(), Irene3000::begin(), CoolBoardLed::blink(), Jetpack::config(), ExternalSensors::config(), CoolWifi::config(), CoolMQTT::config(), CoolTime::config(), Irene3000::config(), CoolBoardSensors::config(), connect(), coolBoardLed, coolBoardSensors, externalSensors, externalSensorsActive, CoolBoardLed::fadeOut(), initReadI2C(), irene3000, ireneActive, jetPack, jetpackActive, mqtt, CoolBoardLed::printConf(), Jetpack::printConf(), ExternalSensors::printConf(), Irene3000::printConf(), CoolMQTT::printConf(), CoolTime::printConf(), CoolBoardSensors::printConf(), CoolWifi::printConf(), rtc, wifiManager, and CoolBoardLed::write()\&.
.PP
.nf
60 {
61 
62 #if DEBUG == 1
63 
64     Serial\&.println( F("Starting the CoolBoard  ")  );
65     Serial\&.println( F("Entering CoolBoard\&.begin() ")  );
66     Serial\&.println();
67 #endif  
68     coolBoardLed\&.printConf();
69     delay(100);
70     
71     coolBoardLed\&.write(255,128,0);//orange
72 
73     this->initReadI2C();
74 
75     coolBoardSensors\&.config();
76     coolBoardSensors\&.begin();
77     coolBoardSensors\&.printConf();
78     delay(100);
79     
80     wifiManager\&.config();
81     wifiManager\&.begin();
82     wifiManager\&.printConf();
83     delay(100);
84 
85     mqtt\&.config();
86     mqtt\&.begin();
87     mqtt\&.printConf();
88     delay(100);
89 
90     if (jetpackActive)
91     {
92         jetPack\&.config();
93         jetPack\&.begin();
94         jetPack\&.printConf();
95         delay(100);
96     }
97 
98     if (ireneActive)
99     {
100         irene3000\&.config();
101         irene3000\&.begin();
102         irene3000\&.printConf();
103         delay(100);
104     }
105 
106     if (externalSensorsActive)
107     {
108         externalSensors\&.config();
109         externalSensors\&.begin();
110         externalSensors\&.printConf();
111         delay(100);
112     }
113     
114     coolBoardLed\&.fadeOut(255,128,0,0\&.5);//orange
115 
116     this->connect();
117     delay(100);
118 
119     rtc\&.config();
120     rtc\&.begin();
121     rtc\&.printConf();
122     delay(100);
123     
124     coolBoardLed\&.blink(0,255,0,0\&.5);//green
125 
126 }
.fi
.SS "bool CoolBoard::config ()"
\fBCoolBoard::config()\fP: This method is provided to configure the \fBCoolBoard\fP : -log interval -irene3000 activated/deactivated -jetpack activated/deactivated -external Sensors activated/deactivated -mqtt server timeout
.PP
\fBReturns:\fP
.RS 4
true if configuration is done, false otherwise 
.RE
.PP

.PP
Definition at line 619 of file CoolBoard\&.cpp\&.
.PP
References CoolBoardLed::begin(), CoolFileSystem::begin(), CoolBoardLed::blink(), CoolBoardLed::config(), coolBoardLed, externalSensorsActive, CoolBoardLed::fadeIn(), CoolBoardLed::fadeOut(), fileSystem, ireneActive, jetpackActive, logInterval, serverTimeOut, sleepActive, CoolBoardLed::strobe(), and userActive\&.
.PP
.nf
620 {
621 
622 #if DEBUG == 1
623 
624     Serial\&.println( F("Entering CoolBoard\&.config() ") );
625     Serial\&.println();
626 
627 #endif
628 
629     //open file system
630     fileSystem\&.begin();
631     
632     //start the led
633     coolBoardLed\&.config();
634     coolBoardLed\&.begin();
635     coolBoardLed\&.fadeIn(243,171,46,0\&.5);//shade of orange     
636 
637     
638     //open configuration file
639     File configFile = SPIFFS\&.open("/coolBoardConfig\&.json", "r");
640     
641     if (!configFile)
642 
643     {
644     
645     #if DEBUG == 1
646 
647         Serial\&.println( F("failed to read /coolBoardConfig\&.json  ") );
648 
649     #endif
650         coolBoardLed\&.blink(255,0,0,0\&.5);//shade of red        
651         return(false);
652     }
653 
654     else
655     {
656         size_t size = configFile\&.size();
657 
658         // Allocate a buffer to store contents of the file\&.
659         std::unique_ptr < char[] > buf(new char[size]);
660 
661         configFile\&.readBytes(buf\&.get(), size);
662 
663         DynamicJsonBuffer jsonBuffer;
664 
665         JsonObject & json = jsonBuffer\&.parseObject(buf\&.get());
666 
667         if (!json\&.success())
668         {
669         
670         #if DEBUG == 1
671 
672             Serial\&.println( F("failed to parse CoolBoard Config json object ") );
673     
674         #endif
675             coolBoardLed\&.blink(255,0,0,0\&.5);//shade of red        
676             return(false);
677         }
678 
679         else
680         {   
681         
682         #if DEBUG == 1
683             
684             Serial\&.println( F("configuration json : ") );
685             json\&.printTo(Serial);
686             Serial\&.println();
687             
688             Serial\&.print(F("jsonBuffer size : "));
689             Serial\&.print(jsonBuffer\&.size());
690             Serial\&.println();
691 
692         #endif
693             
694             //parsing userActive Key
695             if (json["userActive"]\&.success())
696             {
697                 this -> userActive = json["userActive"];
698             }
699 
700             else
701             {
702                 this -> userActive = this -> userActive;
703             }
704             json["userActive"] = this -> userActive;
705 
706             //parsing logInterval key
707             if (json["logInterval"]\&.success())
708             {

src/CoolBoard.cpp
/**
*	\file	CoolBoard.cpp
*  	\brief	CoolBoard Source file
*	\author	Mehdi Zemzem
*	\version 1.0
*  	\date	27/06/2017
*
*
*/

#include "FS.h"
#include "CoolBoard.h"
#include "ArduinoJson.h"
#include "Arduino.h"
#include <Wire.h>
#include <memory>

#define DEBUG 1




/**
*	CoolBoard::CoolBoard():
*	This Constructor is provided to start
*	the I2C interface and Init the different
*	used pins
*/
CoolBoard::CoolBoard()
{

#if DEBUG == 1

	Serial.println( F("Entering CoolBoard Constructor") );
	Serial.println();

#endif
	
	Wire.begin(2, 14);                       //I2C init 

	pinMode(EnI2C, OUTPUT);		   //Declare I2C Enable pin 

}


/**
*	CoolBoard::begin():
*	This method is provided to configure and
*	start the used CoolKit Parts.
*	It also starts the first connection try
*	If Serial is enabled,it prints the configuration
*	of the used parts.
*/
void CoolBoard::begin()
{

#if DEBUG == 1

	Serial.println( F("Starting the CoolBoard  ")  );
	Serial.println( F("Entering CoolBoard.begin() ")  );
	Serial.println();
#endif	

#if DEBUG == 0
	Serial.println( F("Starting Coolboard..."));
#endif


	delay(100);
	
	coolBoardLed.write(255,128,0);//orange

	this->initReadI2C();
	delay(50);

	coolBoardSensors.config();
	coolBoardSensors.begin();
	delay(100);
	
	wifiManager.config();
	wifiManager.begin();
	delay(100);

	mqtt.config();
	mqtt.begin();
	delay(100);

#if DEBUG == 1

	coolBoardLed.printConf();
	coolBoardSensors.printConf();
	wifiManager.printConf();
	mqtt.printConf();

#endif


	if (jetpackActive)
	{
		jetPack.config();
		jetPack.begin();
		#if DEBUG == 1
			jetPack.printConf();
		#endif
		delay(100);
	}

	if (ireneActive)
	{
		irene3000.config();
		irene3000.begin();
		#if DEBUG == 1
			irene3000.printConf();
		#endif
		delay(100);
	}

	if (externalSensorsActive)
	{
		externalSensors.config();
		externalSensors.begin();
		#if DEBUG == 1
			externalSensors.printConf();
		#endif
		delay(100);
	}
	
	coolBoardLed.fadeOut(255,128,0,0.5);//orange

	this->connect();
	delay(100);

	rtc.config();
	rtc.begin();
	#if DEBUG == 1
		rtc.printConf();
	#endif
	delay(100);
	
	coolBoardLed.blink(0,255,0,0.5);//green

}

/**
*	CoolBoard::isConnected()
*	
*	This method is provided to check
*	if the card is connected to Wifi and MQTT
*
*	\return	 0 : connected
*		-1: Wifi Not Connected
*		-2: MQTT Not Connected
*			
*/
int CoolBoard::isConnected()
{

#if DEBUG == 1	

	Serial.println( F("Entering CoolBoard.isConnected ") );
	Serial.println();

#endif
	if (wifiManager.state() != WL_CONNECTED)
	{
	
		Serial.println(F("Wifi Not Connected"));

	#if DEBUG == 1

		Serial.println(F("Wifi State is "));
		Serial.println(wifiManager.state());
		
	#endif
		return(-1);
	}
	
	if(mqtt.state() != 0)
	{
		
		Serial.println( F("MQTT not Connected"));

	#if DEBUG==1
		Serial.println( F("mqtt state is :") );
		Serial.println(mqtt.state());	
	
	#endif

	}
	
	return(0);

}


/**
*	CoolBoard::connect():
*	This method is provided to manage the network
*	connection and the mqtt connection.
*
*	 \return mqtt client state
*/
int CoolBoard::connect()
{

#if DEBUG == 1	

	Serial.println( F("Entering CoolBoard.connect ") );
	Serial.println();
	Serial.println( F("Connecting the CoolBoard  ") );
	delay(100);

#endif
	coolBoardLed.write(0,0,255);//blue

	
			
	
#if DEBUG == 1		

	Serial.println( F("Launching CoolWifi") );
	Serial.println();

#endif
	wifiManager.connect();
	delay(100);


	//only attempt MQTT connection when Wifi is Connected
	if (wifiManager.state() == WL_CONNECTED)
	{

	#if DEBUG == 1	
	
		Serial.println( F("Launching mqtt.connect()") );
		Serial.println();
	
	#endif	
		//logInterval in seconds
		mqtt.connect(this -> getLogInterval());
		delay(100);
	}
	
		
	
	
#if DEBUG == 1

	Serial.println( F("mqtt state is :") );
	Serial.println(mqtt.state());
	Serial.println();
	delay(100);

#endif

	coolBoardLed.blink(0,0,255,0.5);//blue

	return(mqtt.state());
}

/**
*	CoolBoard::onLineMode():
*	This method is provided to manage the online
*	mode:	-update clock
*		-read sensor
*		-do actions
*		-publish data
*		-read answer
*		-update config
*/
void CoolBoard::onLineMode()
{

	coolBoardLed.fadeIn(128,255,50,0.5);//shade of green

#if DEBUG == 1

	Serial.println( F("Entering CoolBoard.onLineMode() ") );
	Serial.println();

#endif
#if DEBUG == 0

	Serial.println( F("CoolBoard is in Online Mode"));

#endif

	data="";
	answer="";

	//send saved data if any
	if(fileSystem.isDataSaved())
	{

		coolBoardLed.fadeIn(128,128,255,0.5);//shade of blue

		Serial.println( F("There is data saved on the File System") );
		Serial.println( F("Sending saved data over MQTT ") );
		Serial.println();
		coolBoardLed.strobe(128,128,255,0.5);//shade of blue 

		mqtt.publish("sending saved data");
		mqtt.mqttLoop();


		
		int size=0;
		std::unique_ptr<String[]> savedData(std::move(fileSystem.getSensorSavedData(size)));//{..,..,..}

		int i=0;
		//loop through the array
		while(i<size)
		{
			//formatting data:
		
			String jsonData = "{\"state\":{\"reported\":";
			jsonData += savedData[i]; // {"state":{"reported":{..,..,..,..,..,..,..,..}
			jsonData += " } }"; // {"state":{"reported":{..,..,..,..,..,..,..,..}  } }

		#if DEBUG == 1 
			Serial.println(F("Size is : "));
			Serial.println(size);
			Serial.print(F("sending line N°"));
			Serial.println(i);
			Serial.println(jsonData);
			Serial.println();

		#endif

			coolBoardLed.strobe(128,128,255,0.5);//shade of blue
		
			mqtt.publish( jsonData.c_str() );
			mqtt.mqttLoop();
		
			coolBoardLed.fadeOut(128,128,255,0.5);//shade of blue
			
			i++;
			yield();
		}		


	#if DEBUG == 1

		Serial.println( F("Saved data sent ") );
		Serial.println();
	
	#endif

	}

	coolBoardLed.blink(128,255,50,0.5);//shade of green

	//clock update
	Serial.println( F("Re-checking RTC..."));
	rtc.update();

	//read user data if user is active
	if(userActive)
	{
		coolBoardLed.fadeIn(245,237,27,0.5);//shade of yellow
	
	#if DEBUG == 1

		Serial.println( F("User is Active") );
		Serial.println( F("Collecting User's data ( mac,username,timeStamp )") );
		Serial.println();
	
	#endif	
		coolBoardLed.blink(245,237,27,0.5);//shade of yellow	

		//reading user data
		data=this->userData();//{"":"","":"","",""}

		//formatting json 
		data.setCharAt( data.lastIndexOf('}') , ',');//{"":"","":"","","",
				
		//read sensors data
	#if DEBUG == 1

		Serial.println( F("Collecting sensors data ") );
		Serial.println();
	
	#endif

		data+=this->readSensors();//{"":"","":"","","",{.......}		

		//formatting json correctly
		data.remove(data.lastIndexOf('{'), 1);//{"":"","":"","","",.......}
		
		coolBoardLed.fadeOut(245,237,27,0.5);//shade of yellow
				
	}	
	else
	{
		//read sensors data
	#if DEBUG == 1

		Serial.println( F("Collecting sensors data ") );
		Serial.println();
	
	#endif
		coolBoardLed.fade(190,100,150,0.5);//shade of violet		
		data=this->readSensors();//{..,..,..}
	}
	
	//do action
	if (jetpackActive)
	{


	#if DEBUG ==1

		Serial.println( F("jetpack is Active ") );
		Serial.println( F("jetpack doing action ") );
		Serial.println();

	#endif
		coolBoardLed.fade(100,100,150,0.5);//dark shade of blue		
		jetPack.doAction(data.c_str());
	}
	
	coolBoardLed.fadeIn(128,255,50,0.5);//shade of green

	//formatting data:
	String jsonData = "{\"state\":{\"reported\":";
	jsonData += data; // {"state":{"reported":{..,..,..,..,..,..,..,..}
	jsonData += " } }"; // {"state":{"reported":{..,..,..,..,..,..,..,..}  } }
	
	//mqtt client loop to allow data handling
	mqtt.mqttLoop();

	coolBoardLed.blink(128,255,50,0.5);//shade of green	

	//read mqtt answer
	answer = mqtt.read();

#if DEBUG == 1 

	Serial.println( F("checking if there's an MQTT message ")  );
	Serial.println( F("answer is : ") );	
	Serial.println(answer);	
	Serial.println();

#endif	

	coolBoardLed.fadeOut(128,255,50,0.5);//shade of green	

	//check if the configuration needs update 
	//and update it if needed 
	this -> update(answer.c_str());
	
	coolBoardLed.fadeIn(128,255,50,0.5);//shade of green	

	//publishing data	
	if( this->sleepActive==0 )	
	{	
		coolBoardLed.strobe(255,0,230,0.5);//shade of pink
		
		//logInterval in seconds
		mqtt.publish( jsonData.c_str(), this->getLogInterval() );
		mqtt.mqttLoop();
	
	}
	else
	{
		coolBoardLed.strobe(230,255,0,0.5);//shade of yellow	

		mqtt.publish(jsonData.c_str());		
		mqtt.mqttLoop();
		answer = mqtt.read();
		this ->update(answer.c_str());

		//logInterval in seconds
		this->sleep( this->getLogInterval() ) ;
	}

	coolBoardLed.fadeOut(128,255,50,0.5);//shade of green		

	mqtt.mqttLoop();

	//read mqtt answer
	answer = mqtt.read();
	this -> update(answer.c_str());	

	coolBoardLed.blink(128,255,50,0.5);//shade of green	


}


/**
*	CoolBoard::offlineMode():
*	This method is provided to manage the offLine
*	mode:	-read sensors
*		-do actions
*		-save data in the file system
*		-if there is WiFi but no Internet : make data available over AP
*		-if there is no connection : retry to connect
*/
void CoolBoard::offLineMode()
{
	coolBoardLed.fade(51,100,50,0.5);//dark shade of green	
#if DEBUG == 1	
	
	Serial.println( F("Entering off line mode ") );	
	
#endif

#if DEBUG == 0

	Serial.println( F("CoolBoard is in Offline Mode"));

#endif

	//read user data if user is active
	if(userActive)
	{

		coolBoardLed.fadeIn(245,237,27,0.5);//shade of yellow

	#if DEBUG == 1
		
		Serial.println( F("User is Active") );
		Serial.println( F("Collecting User's data ( mac,username,timeStamp )") );
		Serial.println();

	#endif

		coolBoardLed.blink(245,237,27,0.5);//shade of yellow	

		//reading user data
		data=this->userData();//{"":"","":"","",""}

		//formatting json 
		data.setCharAt( data.lastIndexOf('}') , ',');//{"":"","":"","","",
		
				
		//read sensors data

		Serial.println( F("Collecting sensors data ") );
		Serial.println();

		data+=this->readSensors();//{"":"","":"","","",{.......}

		

		//formatting json correctly
		data.remove(data.lastIndexOf('{'), 1);//{"":"","":"","","",.......}

		coolBoardLed.fadeOut(245,237,27,0.5);//shade of yellow
				
	}	
	else
	{
		//read sensors data
	#if DEBUG == 1

		Serial.println( F("Collecting sensors data ") );
		Serial.println();

	#endif

		coolBoardLed.fade(190,100,150,0.5);//shade of violet		

		data=this->readSensors();//{..,..,..}
	}

	coolBoardLed.fade(51,100,50,0.5);//dark shade of green	

	//do action
	if (jetpackActive)
	{
	


	#if DEBUG == 1

		Serial.println( F("jetpack is Active ") );
		Serial.println( F("jetpack doing action ") );
		Serial.println();
	
	#endif
		coolBoardLed.fade(100,100,150,0.5);//dark shade of blue	
	
		jetPack.doAction( data.c_str() );
	}
	
	coolBoardLed.fade(51,100,50,0.5);//dark shade of green	
	
	//saving data in the file system
	
	fileSystem.saveSensorData( data.c_str() );

	#if DEBUG == 0

		Serial.println( F("saving Data in Memory : OK"));

	#endif

	coolBoardLed.fadeOut(51,100,50,0.5);//dark shade of green

	//case we have wifi but no internet
	if( (wifiManager.state() == WL_CONNECTED) && ( mqtt.state()!=0 ) )
	{
		
		Serial.println(F("there is Wifi but no Internet"));
		Serial.println(F("lunching AP to check saved files"));
		Serial.println(F("and Add new WiFi if needed"));
		
		wifiManager.connectAP();
		
	}
	
	//case we have no connection at all
	if( wifiManager.state() != WL_CONNECTED )
	{
	
	#if DEBUG == 1
		
		Serial.println(F("there is No Wifi "));
		Serial.println(F("retrying to connect"));
	
	#endif

	#if DEBUG == 0
		Serial.println( F("there is no WiFi..."));
	#endif
		
		this->connect();//nomad case : just run wifiMulti
				//normal case : run wifiMulti+AP
		
	}	

}

/**
*	CoolBoard::config():
*	This method is provided to configure
*	the CoolBoard :	-log interval
*			-irene3000 activated/deactivated
*			-jetpack activated/deactivated
*			-external Sensors activated/deactivated
*			-mqtt server timeout
*
*	\return true if configuration is done,
*	false otherwise
*/
bool CoolBoard::config()
{

#if DEBUG == 1

	Serial.println( F("Entering CoolBoard.config() ") );
	Serial.println();

#endif
#if DEBUG == 0
	Serial.println();
	Serial.println( F("Loading configuration for this CoolBoard..."));
#endif 

	//open file system
	fileSystem.begin();
	
	//start the led
	coolBoardLed.config();
	coolBoardLed.begin();
	coolBoardLed.fadeIn(243,171,46,0.5);//shade of orange		

	
	//open configuration file
	File configFile = SPIFFS.open("/coolBoardConfig.json", "r");
	
	if (!configFile)

	{
	
		Serial.println( F("failed to read /coolBoardConfig.json  ") );

		coolBoardLed.blink(255,0,0,0.5);//shade of red		
		return(false);
	}

	else
	{
		size_t size = configFile.size();

		// Allocate a buffer to store contents of the file.
		std::unique_ptr < char[] > buf(new char[size]);

		configFile.readBytes(buf.get(), size);

		DynamicJsonBuffer jsonBuffer;

		JsonObject & json = jsonBuffer.parseObject(buf.get());

		if (!json.success())
		{
		
			Serial.println( F("failed to parse CoolBoard Config json object ") );
	
			coolBoardLed.blink(255,0,0,0.5);//shade of red		
			return(false);
		}

		else
		{	
		
		#if DEBUG == 1
			
			Serial.println( F("configuration json : ") );
			json.printTo(Serial);
			Serial.println();
			
			Serial.print(F("jsonBuffer size : "));
			Serial.print(jsonBuffer.size());
			Serial.println();

		#endif
			
			//parsing userActive Key
			if (json["userActive"].success())
			{
				this -> userActive = json["userActive"];
			}

			else
			{
				this -> userActive = this -> userActive;
			}
			json["userActive"] = this -> userActive;

			//parsing logInterval key
			if (json["logInterval"].success())
			{
				this -> logInterval = json["logInterval"];
			}
			else
			{
				this -> logInterval = this -> logInterval;
			}
			json["logInterval"] = this -> logInterval;
			
			//parsing ireneActive key			
			if (json["ireneActive"].success())
			{
				this -> ireneActive = json["ireneActive"];
			}
			else
			{
				this -> ireneActive = this -> ireneActive;
			}
			json["ireneActive"] = this -> ireneActive;
			
			//parsing jetpackActive key
			if (json["jetpackActive"].success())
			{
				this -> jetpackActive = json["jetpackActive"];
			}
			else
			{
				this -> jetpackActive = this -> jetpackActive;
			}
			json["jetpackActive"] = this -> jetpackActive;

			//parsing externalSensorsActive key
			if (json["externalSensorsActive"].success())
			{
				this -> externalSensorsActive = json["externalSensorsActive"];
			}
			else
			{
				this -> externalSensorsActive = this -> externalSensorsActive;
			}
			json["externalSensorsActive"] = this -> externalSensorsActive;

			
			//parsing sleepActive key
			if (json["sleepActive"].success())
			{
				this -> sleepActive = json["sleepActive"];
			}
			else
			{
				this -> sleepActive = this -> sleepActive;
			}
			json["sleepActive"] = this -> sleepActive;

			//saving the current/correct configuration
			configFile.close();
			configFile = SPIFFS.open("/coolBoardConfig.json", "w");
			if (!configFile)
			{
			
				Serial.println( F("failed to write to /coolBoardConfig.json") );
				Serial.println();

 				coolBoardLed.blink(255,0,0,0.5);//shade of red		
				return(false);
			}

			json.printTo(configFile);
			configFile.close();
			#if DEBUG == 0

				Serial.println( F("Configuration loaded : OK"));
				Serial.println();

			#endif

			return(true);
		}
	}

	coolBoardLed.strobe(243,171,46,0.5);//shade of orange
	
	coolBoardLed.fadeOut(243,171,46,0.5);//shade of orange				
}

/**
*
*	CoolBoard::printConf():
*	This method is provided to print
*	the configuration to the Serial
*	Monitor.
*/
void CoolBoard::printConf()
{

#if DEBUG == 1
	
	Serial.println( F("Entering CoolBoard.printConf() ") );
	Serial.println();

#endif

	Serial.println( F("Printing Cool Board Configuration "));
	Serial.print( F("log interval 		: "));
	Serial.println(this->logInterval);

	Serial.print( F("irene active 		: "));
	Serial.println(this->ireneActive);

	Serial.print( F("jetpack active		: "));
	Serial.println(this->jetpackActive);

	Serial.print( F("external sensors active 	: "));
	Serial.println(this->externalSensorsActive);

	Serial.print( F("sleept active 		: "));
	Serial.println(this->sleepActive);

	Serial.print( F("user active 		: "));
	Serial.println(this->userActive);

	Serial.println();




}

/**
*	CoolBoard::update(mqtt answer):
*	This method is provided to handle the
*	configuration update of the different parts
*/
void CoolBoard::update(const char * answer)
{
	coolBoardLed.fadeIn(153,76,0,0.5);//shade of brown		

#if DEBUG == 1

	Serial.println( F("Entering CoolBoard.update() ") );
	Serial.println();
	Serial.println( F("message is : ") );
	Serial.println(answer);
	Serial.println();

#endif

	DynamicJsonBuffer jsonBuffer;
	JsonObject & root = jsonBuffer.parseObject(answer);
	JsonObject & stateDesired = root["state"];

#if DEBUG == 1

	Serial.println( F("root json : ") );
	root.printTo(Serial);
	Serial.println();

	Serial.println( F("stateDesired json : "));
	stateDesired.printTo(Serial);
	Serial.println();
	
	Serial.print( F("jsonBuffer size : "));
	Serial.println(jsonBuffer.size());

#endif

	if (stateDesired.success())
	{
	
	#if DEBUG == 1

		Serial.println( F("update message parsing : success") );
		Serial.println();
	
	#endif

			String answerDesired;
		
			stateDesired.printTo(answerDesired);
			
		#if DEBUG == 1		
		
			Serial.println( F("update is ok ") );
			Serial.println( F("desired update is : ") );			
			Serial.println(answerDesired);
			Serial.println("json size is : ");
			Serial.println(jsonBuffer.size() ) ;				
			Serial.println();

		
		#endif
			//saving the new configuration
			fileSystem.updateConfigFiles(answerDesired);

		        //answering the update msg:
			//reported = received configuration
			//desired=null
		
			String updateAnswer;
			String tempString;
			
			stateDesired.printTo(tempString);
			updateAnswer="{\"state\":{\"reported\":";
			updateAnswer+=tempString;
			updateAnswer+=",\"desired\":null}}";

		#if DEBUG == 1

			Serial.println( F("preparing answer message ") );
			Serial.println();
			Serial.println( F("updateAnswer : ") );
			Serial.println(updateAnswer);
		
		#endif	

			mqtt.publish(updateAnswer.c_str());
			
			mqtt.mqttLoop();

			delay(10);
			
			//restart the esp to apply the config
			ESP.restart();
	}
	else
	{
	
	#if DEBUG == 1

		Serial.println( F("Failed to parse update message( OR no message received )") );
		Serial.println();
	
	#endif
	
	}

	coolBoardLed.strobe(153,76,0,0.5);//shade of brown
	coolBoardLed.fadeOut(153,76,0,0.5);//shade of brown								
}

/**
*	CoolBoard::getLogInterval():
*	This method is provided to get
*	the log interval
*
*	\return interval value in s
*/
unsigned long CoolBoard::getLogInterval()
{

#if DEBUG == 1

	Serial.println( F("Entering CoolBoard.getLogInterval() ") );
	Serial.println();
	Serial.println( F("log Interval is :") );
	Serial.println(logInterval);
	Serial.println();

#endif

	return(this -> logInterval);
}

/**
*	CoolBoard::readSensors():
*	This method is provided to read and
*	format all the sensors data in a
*	single json.
*	
*	\return	json string of all the sensors read.
*/
String CoolBoard::readSensors()
{

	coolBoardLed.fadeIn(128,255,0,0.5);//light shade of green
				
#if DEBUG == 1

	Serial.println( F("Entering CoolBoard.readSensors()") );
	Serial.println();

#endif
	coolBoardLed.strobe(128,255,0,0.5);//light shade of green

	String sensorsData;
	
	this->initReadI2C();

	sensorsData = coolBoardSensors.read(); // {..,..,..}
	
	if (externalSensorsActive)
	{
		sensorsData += externalSensors.read(); // {..,..,..}{..,..}

		sensorsData.setCharAt(sensorsData.lastIndexOf('}'), ','); // {..,..,..}{..,..,
		sensorsData.setCharAt(sensorsData.lastIndexOf('{'), ','); // {..,..,..},..,..,
		sensorsData.remove(sensorsData.lastIndexOf('}'), 1); // {..,..,..,..,..,
		sensorsData.setCharAt(sensorsData.lastIndexOf(','), '}'); // {..,..,..,..,..}

	}
	if (ireneActive)
	{
		sensorsData += irene3000.read(); // {..,..,..,..,..}{..,..,..}

		sensorsData.setCharAt(sensorsData.lastIndexOf('}'), ','); // {..,..,..,..,..}{..,..,..,
		sensorsData.setCharAt(sensorsData.lastIndexOf('{'), ','); // {..,..,..,..,..},..,..,..,
		sensorsData.remove(sensorsData.lastIndexOf('}'), 1); // {..,..,..,..,..,..,..,..,
		sensorsData.setCharAt(sensorsData.lastIndexOf(','), '}'); // {..,..,..,..,..,..,..,..}		
		
		
	}

	//getting Hour:
	tmElements_t tm;
	tm=rtc.getTimeDate();
	
	//adding Hour
	sensorsData.remove(sensorsData.lastIndexOf('}'), 1); // {..,..,..,..,..,..,..,..,	
	sensorsData+=",\"hour\":";	
	sensorsData+=tm.Hour;
	sensorsData+="}";
	
#if DEBUG == 1
	Serial.println();
	Serial.println( F("sensors data is ") );
	Serial.println(sensorsData);
	Serial.println();

#endif
	coolBoardLed.fadeOut(128,255,0,0.5);//light shade of green

	return(sensorsData);

}

/**
*	CoolBoard::initReadI2C():
*	This method is provided to enable the I2C
*	Interface.
*/
void CoolBoard::initReadI2C()
{

#if DEBUG == 1

 	Serial.println( F("Entering CoolBoard.initReadI2C()") );
	Serial.println();

#endif
 
	digitalWrite(EnI2C,HIGH);//HIGH= I2C Enable

}


/**
*	CoolBoard::userData():
*	This method is provided to
*	return the user's data.
*	
*	\return json string of the user's data
*/
String CoolBoard::userData()
{

#if DEBUG == 1

	Serial.println( F("Entering CoolBoard.userData() ") );
	Serial.println();

#endif

	String tempMAC = WiFi.macAddress();

	tempMAC.replace(":", "");

	String userJson = "{\"user\":\"";

	userJson += mqtt.getUser();

	userJson += "\",\"timestamp\":\"";

	userJson += rtc.getESDate(); // "timestamp":"20yy-mm-ddThh:mm:ssZ"

	userJson += "\",\"mac\":\"";

	userJson += tempMAC;

	userJson += "\"}";

#if DEBUG == 1

	Serial.println( F("userData is : ") );
	Serial.println(userJson);
	Serial.println();

#endif	
	
	return(userJson);
	
}


/**
*	CoolBoard::sleep(int interval):
*	This method is provided to allow the
*	board to enter deepSleep mode for
*	a period of time equal to interval in s 
*/
void CoolBoard::sleep(unsigned long interval)
{

	Serial.println( F("Entering CoolBoard.sleep() ") );
	Serial.print( F("going to sleep for ") );
	Serial.print(interval);
	Serial.println(F("s") );
	Serial.println();
	
	//interval is in seconds , interval*1000*1000 in µS
	ESP.deepSleep ( ( interval * 1000 * 1000 ), WAKE_RF_DEFAULT) ;

}



